ngraph/core/src/op/matmul.cpp

```diff
@@ -113,68 +53,170 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
+            return PartialShape(Shape{});
         }
-        if (transpose_b && arg1_rank > 1)
+
+        if (transpose_a && arg0_rank > 1)
         {
-            swap(arg1_shape_update[arg1_rank - 2], arg1_shape_update[arg1_rank - 1]);
+            swap(arg0_shape_tmp[arg0_rank - 2], arg0_shape_tmp[arg0_rank - 1]);
         }
 
-        if (arg0_rank == 1 && arg1_rank == 1)
+        if (transpose_b && arg1_rank > 1)
         {
-            NGRAPH_CHECK(arg0_shape_update == arg1_shape_update, "Incompatible arg shapes");
-            output_shape = Shape{};
+            swap(arg1_shape_tmp[arg1_rank - 2], arg1_shape_tmp[arg1_rank - 1]);
         }
-        else if (arg0_rank == 1)
+
+        if (arg0_rank == 1)
         {
-            // i.e., arg0 shape {3}, arg1 shape{2, 3, 2}, output shape {2, 2}
-            NGRAPH_CHECK(arg0_shape_update[0] == arg1_shape_update[arg1_rank - 2],
-                         "Incompatible arg shapes");
-            arg1_shape_update.erase(arg1_shape_update.begin() + arg1_rank - 2);
-            output_shape = arg1_shape_update;
+            // i.e. arg0 shape {3}, arg1 shape{2, 3, 2}, output shape {2, 2}
+            NGRAPH_CHECK(Dimension::merge(merged_dimension,
+                                          arg0_shape_tmp[0],
+                                          arg1_shape_tmp[arg1_shape_tmp.size() - 2]),
+                         "Incompatible matrix dimensions.");
+            arg1_shape_tmp.erase(arg1_shape_tmp.begin() + arg1_rank - 2);
+            return PartialShape(arg1_shape_tmp);
         }
         else if (arg1_rank == 1)
         {
-            // i.e., arg0 shape {2, 2, 3}, arg1 shape{3}, output shape {2, 2}
-            NGRAPH_CHECK(arg1_shape_update[0] == arg0_shape_update[arg0_rank - 1],
-                         "Incompatible arg shapes");
-            arg0_shape_update.erase(arg0_shape_update.begin() + arg0_rank - 1);
-            output_shape = arg0_shape_update;
-        }
-        else if (arg0_rank == 2 && arg1_rank == 2)
-        {
-            NGRAPH_CHECK(arg0_shape_update[1] == arg1_shape_update[0], "Incompatible arg shapes");
-            output_shape = Shape{arg0_shape_update[0], arg1_shape_update[1]};
+            // i.e. arg0 shape {2, 2, 3}, arg1 shape{3}, output shape {2, 2}
+            NGRAPH_CHECK(Dimension::merge(
+                             merged_dimension, arg1_shape_tmp[0], arg0_shape_tmp[arg0_rank - 1]),
+                         "Incompatible matrix dimensions.");
+            arg0_shape_tmp.erase(arg0_shape_tmp.begin() + arg0_rank - 1);
+            return PartialShape(arg0_shape_tmp);
         }
-        else
+
+        // 2D and bigger tensors cases.
+        NGRAPH_CHECK(Dimension::merge(merged_dimension,
+                                      arg0_shape_tmp[arg0_rank - 1],
+                                      arg1_shape_tmp[arg1_rank - 2]),
+                     "Incompatible matrix dimensions.");
```

Comment:
Referring our discussion yesterday when we were not able to decide which behavior we should choose when have two dynamic dimensions that are not really compatible. It is relevant to this place and places below about incompatibilities during the broadcasting of batch dimensions. I mean cases like we trying to match [2,3] and [4,5] for example. I think it should fail (as written in this code, right?). I think we should follow this approach of aggressive failing because the original validation code for MatMul doesn't handle dynamic cases at all except prevalidate function. @mitruska, can you confirm that old MatMul::validate_and_infer_types cannot proceed with dynamic inputs?

Reply:
Old (current master) implementation for dynamic cases preserves only rank.
For the same test case as [matmul_batch_dynamic_bounds](https://github.com/openvinotoolkit/openvino/blob/9c2a38ca338fed06e75a460c696e547d121909ee/ngraph/test/type_prop/matmul.cpp#L433-L474) the output is:
> Expected equality of these values:
  matmul->get_output_partial_shape(0)
    Which is: {?,?,?,?,?,?,?,?,?,?,?,?,?}
  expected_output_shape
    Which is: {?,[10, 20],10,[5, 7],[5, 7],[5, 7],?,[0, 1],[2, 5],?,3,5,4}

Reply:
So it doesn't catch case when dimensions are not compatible and then just set ? for corresponding output dimension?

Reply:
Yes, unfortunately current master doesn't throw with incompatible batch dynamic bounds like: https://github.com/openvinotoolkit/openvino/blob/2c7360938ae35291c76f5bb2ad4d17d3c514579c/ngraph/test/type_prop/matmul.cpp#L552-L571

Reply:
So I can check here https://github.com/openvinotoolkit/openvino/blob/2c7360938ae35291c76f5bb2ad4d17d3c514579c/ngraph/core/src/op/matmul.cpp#L135-L139 returned value from `Dimension::merge` and throw only for static, otherwise set `Dimensions::dynamic()`

Reply:
Looks like we have to implement the same logic for another Dimension::merge call below for other dimensions. I would propose to introduce a helper function for this backward compatible goal which should be used instead of raw Dimension::merge.